gennet-module/gennet.py

```diff
@@ -265,103 +247,79 @@ def generate_and_write_files(dirname, num_topics, num_subnets, node_type_distrib
                     sys.exit(1)
+                if "general" in conf and "prng_seed" in conf["general"]:
+                    conf["gennet"]["prng_seed"] = conf["general"]["prng_seed"]
                 # TODO : type-check and sanity-check the config.json
                 #print(conf)
+            ctx.default_map.update(conf["gennet"])  # Merge config and default_map
         except Exception as ex:
             raise typer.BadParameter(str(ex))
     return cfile
 
 
-# methods to merge cli values, config.json and default values
-def test_and_set_int(cli_val, file_val, conf, module="gennet"):
-    if cli_val != INT_NONE:
-        return cli_val
-    if cli_val == INT_NONE and module in conf and file_val in conf[module]:
-        return conf[module][file_val]
-    return defaults[file_val]
+# sanity check : num_partitions == 1
+def _num_partitions_callback(num_partitions: int):
+    if num_partitions > 1:
+        raise ValueError(
+            f"--num-partitions {num_partitions}, Sorry, we do not yet support partitions")
+    return num_partitions
 
 
-def test_and_set_str(cli_val, file_val, conf, module="gennet"):
-    if cli_val != STR_NONE:
-        return cli_val
-    if cli_val == STR_NONE and module in conf and file_val in conf[module]:
-        return conf[module][file_val]
-    return defaults[file_val]
+# sanity check :  num_subnets < num_nodes
+def _num_subnets_callback(ctx: typer, Context, num_subnets: int):
+    num_nodes = ctx.params["num_nodes"]
+    if num_subnets == -1:
+        num_subnets = num_nodes
+    if num_subnets > num_nodes:
+        raise ValueError(
+            f"num_subnets must be <= num_nodes: num_subnets={num_subnets}, num_nodes={1}")
+    return num_subnets
 
 
-def main(benchmark : bool = False,
-         output_dir: str = STR_NONE,
-         prng_seed: int = INT_NONE,
-         num_nodes: int = INT_NONE,
-         num_topics: int = INT_NONE,
+def main(ctx: typer.Context, benchmark : bool = False,
+         output_dir: str = "network_data",
+         prng_seed: int = 3,
+         num_nodes: int = 4,
+         num_topics: int = 1,
          network_type: networkType = networkType.NEWMANWATTSSTROGATZ.value,
-         num_partitions: int = typer.Option(INT_NONE, callback=_num_partitions_callback),
-         num_subnets: int = typer.Option(INT_NONE, callback=_num_subnets_callback),
-         config_file: str = typer.Option(STR_NONE, callback=conf_callback, is_eager=True)):
+         num_subnets: int = typer.Option(1, callback=_num_subnets_callback),
+         num_partitions: int = typer.Option(1, callback=_num_partitions_callback),
+         config_file: str = typer.Option("", callback=conf_callback, is_eager=True)):
 
     # Benchmarking: record start time and start tracing mallocs
     if benchmark :
         start = time.time()
         tracemalloc.start()
 
-
+    # re-read the conf file to set node_type_distribution -- no cli equivalent
```

Comment:
I totally missed this before. if you want to set a json as parameter there are multiple [workarounds](https://github.com/tiangolo/typer/issues/219) which I think they will be more maintainable.

Reply:
the real solution is to add  node type distribution as a cli argument and am currently deciding on the minimal way to input a dict from cli. this snippet is a stop-gap and will disappear once I do it in the next PR.

Reply:
Ye ye but there is ways like the link I posted here where it can be implemented, but I guess it can be done in a future PR.